File: 1370-count-number-of-nice-subarrays/count-number-of-nice-subarrays.java
import java.util.HashMap;

class Solution {
    public int numberOfSubarrays(int[] nums, int k) {
        HashMap<Integer, Integer> countMap = new HashMap<>();
        countMap.put(0, 1); // base case: one way to have 0 odd numbers with an empty subarray
        int oddCount = 0;
        int result = 0;

        for (int num : nums) {
            if (num % 2 != 0) {
                oddCount++;
            }
            if (countMap.containsKey(oddCount - k)) {
                result += countMap.get(oddCount - k);
            }
            countMap.put(oddCount, countMap.getOrDefault(oddCount, 0) + 1);
        }

        return result;
    }
}

File: 1370-count-number-of-nice-subarrays/count-number-of-nice-subarrays.cpp
class Solution {
public:
    int helper(vector<int>& nums, int goal){
        if(goal<0) return 0;
        int count = 0;
        int sum = 0;
        int left = 0;
        int right = 0;
        while(right<nums.size()){
            sum+=(nums[right])%2;
            while(left<=right && sum>goal){
                sum = sum-(nums[left]%2);
                left++;
            }
            count+=right-left+1;
            right++;
        }
        return count;
    }
    int numberOfSubarrays(vector<int>& nums, int k) {
        return helper(nums,k)-helper(nums,k-1);
    }
};